### PR TITLE
[BFT] `model.Proposal` refactoring

### PR DESCRIPTION
**Pull request by durkmurder**
https://github.com/onflow/flow-go/issues/6517

### Context 

This PR implements refactoring of `model.Proposal`, all details are in attached issue. This PR very closes follows proposed solution.

### CODE REVIEW

**Review by jordanschalm — APPROVED** [review]

File: consensus/hotstuff/safetyrules/safety_rules.go [hunk 1]
```diff
@@ -153,6 +172,7 @@ func (r *SafetyRules) ProduceVote(proposal *model.Proposal, curView uint64) (*mo
 	}
 
 	return vote, nil
+
```
Comment (AlexHentschel) [inline consensus/hotstuff/safetyrules/safety_rules.go:175]:
```suggestion
```

File: consensus/hotstuff/verification/combined_signer_v2_test.go [hunk 2]
```diff
@@ -78,7 +78,7 @@ func TestCombinedSignWithBeaconKey(t *testing.T) {
 	// check that a created proposal can be verified by a verifier
```
Comment (AlexHentschel) [inline consensus/hotstuff/verification/combined_signer_v2_test.go:78]:
```suggestion
	// check that the proposer's vote for their own block (i.e. the proposer signature in the header) passes verification
```

File: consensus/hotstuff/verification/combined_signer_v2_test.go [hunk 3]
```diff
@@ -190,7 +190,6 @@ func TestCombinedSignWithNoBeaconKey(t *testing.T) {
...
```
Comment (AlexHentschel) [inline consensus/hotstuff/verification/combined_signer_v2_test.go:190]:
```suggestion
	// check that the proposer's vote for their own block (i.e. the proposer signature in the header) passes verification
```

File: consensus/hotstuff/verification/combined_signer_v3_test.go [hunk 4]
```diff
@@ -58,11 +58,10 @@ func TestCombinedSignWithBeaconKeyV3(t *testing.T) {
 	verifier := NewCombinedVerifierV3(committee, packer)
 
 	// check that a created proposal can be verified by a verifier
```
Comment (AlexHentschel) [inline consensus/hotstuff/verification/combined_signer_v3_test.go:60]:
```suggestion
	// check that the proposer's vote for their own block (i.e. the proposer signature in the header) passes verification
```

File: consensus/hotstuff/verification/combined_signer_v3_test.go [hunk 5]
```diff
@@ -121,11 +120,10 @@ func TestCombinedSignWithNoBeaconKeyV3(t *testing.T) {
 	packer := signature.NewConsensusSigDataPacker(committee)
 	verifier := NewCombinedVerifierV3(committee, packer)
 
-	proposal, err := signer.CreateProposal(block)
+	vote, err := signer.CreateVote(block)
```
Comment (AlexHentschel) [inline consensus/hotstuff/verification/combined_signer_v3_test.go:123]:
```suggestion
	// check that the proposer's vote for their own block (i.e. the proposer signature in the header) passes verification
	vote, err := signer.CreateVote(block)
```

File: consensus/hotstuff/votecollector/combined_vote_processor_v3_test.go [hunk 6]
```diff
@@ -982,8 +982,7 @@ func TestCombinedVoteProcessorV3_BuildVerifyQC(t *testing.T) {
 
 	leader := stakingSigners[0]
```
Comment (AlexHentschel) [inline consensus/hotstuff/votecollector/combined_vote_processor_v3_test.go:984]:
```suggestion
```

File: consensus/hotstuff/votecollector/statemachine_test.go [hunk 7]
```diff
@@ -90,9 +90,9 @@ func (s *StateMachineTestSuite) TestStatus_StateTransitions() {
 	require.Equal(s.T(), hotstuff.VoteCollectorStatusVerifying, s.collector.Status())
 
 	// after submitting double proposal we should transfer into invalid state
-	err = s.collector.ProcessBlock(helper.MakeProposal(
-		helper.WithBlock(
-			helper.MakeBlock(helper.WithBlockView(s.view)))))
+	err = s.collector.ProcessBlock(helper.MakeSignedProposal(helper.WithProposal(
```
Comment (AlexHentschel) [inline consensus/hotstuff/votecollector/statemachine_test.go:93]: can we use the function `makeSignedProposalWithView` here too? Just skip this suggestion if we can't. No detailed explanation needed.

**Review by AlexHentschel — APPROVED** [review]
very cool. Thanks for the cleanup. Minor comments regarding documentation only.

File: consensus/hotstuff/model/proposal.go [hunk 8]
```diff
@@ -4,16 +4,32 @@ import (
...
+// we require the QC (mandatory part of every block) and optionally a TC (of the QC in the block
+// is _not_ for the immediately previous view). Note that the proposer's signature for its own
+// block is _not_ required.
+// By explicitly differentiating the Proposal from the SignedProposal (extending Proposal by
+// adding the proposer's signature), we can unify the algorithmic path of signing block proposals.
+// This codifies the important aspect that a proposer's signature for their own block
+// is conceptually also just a vote (we explicitly use that for aggregating votes, including the
+// proposer's own vote to a QC). In order to express this conceptual equivalence in code, the
+// voting logic in Safety Rules must also operate on an unsigned Proposal.
 type Proposal struct {
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply consensus/hotstuff/model/proposal.go:29]: revised in https://github.com/onflow/flow-go/pull/6526/commits/497654848c0e13748ed547db62986f752b29f5e5

**Pull request merged**